HealthVault/Classes/Caching/Private/MHVThingCacheDatabase.m

```diff
@@ -94,145 +99,295 @@ - (MHVCachedThing *)newThingForRecord:(MHVCachedRecord *)record
+    }
+    if ([NSArray isNilOrEmpty:thingIds])
+    {
+        //Nothing to delete
+        if (completion)
         {
-            //Nothing to delete
-            return nil;
+            completion(nil);
         }
-        
-        if ([NSString isNilOrEmpty:recordId])
+        return;
+    }
+    
+    if ([NSString isNilOrEmpty:recordId])
+    {
+        if (completion)
         {
-            return [NSError MVHRequiredParameterIsNil];
+            completion([NSError MVHRequiredParameterIsNil]);
         }
-        
-        __block NSError *error;
-        
-        [self.managedObjectContext performBlockAndWait:^
+        return;
+    }
+    
+    [self.managedObjectContext performBlock:^
+     {
+         NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"MHVCachedThing"];
+         
+         NSMutableArray<NSString *> *where = [NSMutableArray new];
+         for (NSString *thingId in thingIds)
          {
-             NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"MHVCachedThing"];
-             
-             NSMutableArray<NSString *> *where = [NSMutableArray new];
-             for (NSString *thingId in thingIds)
+             [where addObject:thingId];
+         }
+         
+         NSPredicate *predicate = [NSPredicate predicateWithFormat:@"thingId IN %@ AND record.recordId == %@", where, [recordId lowercaseString]];
+         [fetchRequest setPredicate:predicate];
+         
+         NSError *error = nil;
+         NSArray *fetchedObjects = [self.managedObjectContext executeFetchRequest:fetchRequest error:&error];
+         for (MHVCachedRecord *record in fetchedObjects)
+         {
+             [self.managedObjectContext deleteObject:record];
+         }
+         
+         error = [self saveContext];
+         
+         if (error)
+         {
+             MHVLOG(@"ThingCacheDatabase: Setting record as invalid, error deleting things %@", error);
+             [self setCacheInvalidForRecordId:recordId
+                                   completion:nil];
+         }
+
+         if (completion)
+         {
+             completion(error);
+         }
+     }];
+}
+
+- (void)cachedResultsForQuery:(MHVThingQuery *)query
+                     recordId:(NSString *)recordId
+                   completion:(void(^)(MHVThingQueryResult *_Nullable queryResult, NSError *_Nullable error))completion
+{
+    NSDate *startDate = [NSDate date];
+    
+    //If no completion, don't need to do the query
+    if (!completion)
+    {
+        return;
+    }
+    
+    @synchronized (self.lockObject)
+    {
+        if (self.deleted)
+        {
+            completion(nil, nil);
+            return;
+        }
+    }
+
+    if (!recordId)
+    {
+        completion(nil, [NSError MVHRequiredParameterIsNil]);
+        return;
+    }
+    
+    MHVCacheQuery *cacheQuery = [[MHVCacheQuery alloc] initWithQuery:query];
+    
+    if (cacheQuery.error)
```

Comment:
This code is redundant, you can just do:

if (!cacheQuery.canQueryCache)
 +    {
 +        completion(nil, cacheQuery.error);
 +        return;
 +    }

The cacheQuery object will provide an error as appropriate, and canQueryCache should always be NO if there is an error.